### PR TITLE
Close the FSA and use dirty trick to unmap it

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli PR

### CODE REVIEW

**Review by vekterli — APPROVED** [review]
This fills me with some dread, but the security implications of this (re: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=4724038) seems primarily to be when a VM is operating in a multi-tenancy context, which shouldn't be the case for users of this class (i.e. a container VM runs code for a single application). Seems paradoxical to have to resort to hacks to keep Java from leaking resources, but here we are.

@bjorncs could you also take a look? Things to keep in back of the mind: deprecations in upcoming Java versions and what might happen if we introduce security managers et al.

**Review by bjorncs — APPROVED** [review]
Looks okay. Afaik there is no other way to trigger unmapping except reflection. This seems to be a limitation with the existing JDK versions.

File: fsa/src/main/java/com/yahoo/fsa/FSA.java [hunk 1]
```diff
@@ -316,13 +325,134 @@ public Iterator iterator(State state) {
...
+                    unsafeClass = Class.forName("jdk.internal.misc.Unsafe");
+                }
+                Method clean = unsafeClass.getMethod("invokeCleaner", ByteBuffer.class);
+                clean.setAccessible(true);
+                Field theUnsafeField = unsafeClass.getDeclaredField("theUnsafe");
+                theUnsafeField.setAccessible(true);
+                Object theUnsafe = theUnsafeField.get(null);
+                clean.invoke(theUnsafe, mmap);
+            } catch (Exception e) {
+                throw new IllegalArgumentException("Falied unmapping ", e);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply fsa/src/main/java/com/yahoo/fsa/FSA.java]: Fixed

File: fsa/src/main/java/com/yahoo/fsa/FSA.java [hunk 2]
```diff
@@ -316,13 +325,134 @@ public Iterator iterator(State state) {
...
+        private final boolean _ok;
+    }
     private final boolean _ok;
-    private final MappedByteBuffer _header;
-    private final MappedByteBuffer _symbol_tab;
-    private final MappedByteBuffer _state_tab;
-    private final MappedByteBuffer _data;
-    private final MappedByteBuffer _phash;
     private final Charset _charset;
+    private final AtomicReference<Maps> maps = new AtomicReference<>();
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply fsa/src/main/java/com/yahoo/fsa/FSA.java]: Time will show....

**Pull request merged**